### PR TITLE
Refactor KPI Forecasting

### PR DESCRIPTION
**Pull request by bochocki**
This is a major refactor that aims to make KPI forecasting easier to modify in the future. This refactor maintains the central design pattern -- forecasts are made from a metric based on configs set in a YAML file -- but most of the implementation has been refactored.

There is a related, ongoing project to [re-model the way we store forecast data in Big Query](https://docs.google.com/document/d/18esfJraogzUf1gbZv25vgXkHigCLefazyvzly9s-1k0/edit#heading=h.3vp2yzc6seen). One key change in this refactor is to format output data to be as close as possible to what we expect the new data model to look like, while also preserving the legacy data model for the time being. Methods related to legacy outputs are clearly marked in the code (usually including `legacy` in the method name) and are commented with `TODO` items that indicate which code paths can be removed once the transition has been made. I tested that the legacy data format was correct by copying legacy tables to `moz-fx-data-bq-data-science.bochocki` and appending output to those tables.

Key changes are:
  - Convert from a functional framework to an object-oriented framework. This will enable us to add new forecasting models more easily in the future by simply inheriting a `BaseForecast` class instead of writing bespoke logic for each new forecasting method.
  - Update the output to produce a single table that can accommodate different numbers of measurements via the `measures` column.
  - Use Metric Hub definitions instead of bespoke metric queries. By using versioned Metric Hub definitions, we can more easily ensure that the forecast is using up-to-date metric values.
 

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is
  referenced, the pull request should include the bug number in the title)
- [ ] Scan the PR and verify that no changes (particularly to
  `.circleci/config.yml`) will cause environment variables (particularly
  credentials) to be exposed in test logs
- [ ] Ensure the container image will be using permissions granted to
  [telemetry-airflow](https://github.com/mozilla/telemetry-airflow/)
  responsibly.

### CODE REVIEW

File: jobs/kpi-forecasting/.gitignore
Comment (kik-kik) [inline jobs/kpi-forecasting/.gitignore:12]: Could you also add `*.pyc` to the ignore list?

File: jobs/kpi-forecasting/Dockerfile
Comment (kik-kik) [inline jobs/kpi-forecasting/Dockerfile:15]: So the best practice would to combine commands into a single layer if possible.

In this case, we could do the following:
```
RUN apt-get -y update && apt-get -y install \
    git \
    && rm -rf /var/lib/apt/lists/*
```

Here's a Stackoverflow comment about this:
https://stackoverflow.com/a/39330458

File: jobs/kpi-forecasting/Dockerfile
Comment (kik-kik) [inline jobs/kpi-forecasting/Dockerfile:17]: For the pip install commands you should also include the `--no-cache-dir` flag. In general in the Dockerfile we try to keep each layer as small as possible.

Can read more about Dockerfile best practices here: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/

File: jobs/kpi-forecasting/README.md
Comment (kik-kik) [inline jobs/kpi-forecasting/README.md:47]: Is the virtual environment created in the project directory by conda? If yes, perhaps we should include it in the `.gitignore`

File: jobs/kpi-forecasting/ci_job.yaml
Comment (kik-kik) [inline jobs/kpi-forecasting/ci_job.yaml:15]:
```suggestion
        command: docker run app:build black .
```

File: jobs/kpi-forecasting/kpi-forecasting/kpi_forecasting.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/kpi_forecasting.py:14]: Is there a need to put YAML and CLI into separate modules? Both appear to be single functions and this appears like a quite a short module.

File: jobs/kpi-forecasting/kpi-forecasting/metric_hub.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/metric_hub.py:31]: some validation on the date fields would be nice since this is user input and we expect a specific format.

File: jobs/kpi-forecasting/kpi-forecasting/metric_hub.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/metric_hub.py:70]: If the enquoting is only for the query then you could just include the `''` inside the query template instead and avoid having to do this additional operation.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:37]: Is this comment providing much value to the readers understanding?

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:62]: Comments should provide understanding not just state what the code does. In this case, it would be more valuable to understand why this is needed. What do we mean by legacy?

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:73]: Is it possible for both `desktop` and `mobile` to be in the set?

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:115]: is this expected and intended behaviour? Or does one of these columns missing a sign of something bad?

File: jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py:22]: Could avoid the if, else if you flipped around the condition `not in` and if True raise the exception. and if we make it through the block do your assignment. Looks like it should always happen.

File: jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py:25]: Is ValueError the appropriate Exception type for this error?

File: jobs/kpi-forecasting/requirements.txt
Comment (kik-kik) [inline jobs/kpi-forecasting/requirements.txt:16]: Are all of these requirements necessary?

File: jobs/kpi-forecasting/kpi-forecasting/metric_hub.py
Comment (kik-kik) [inline jobs/kpi-forecasting/kpi-forecasting/metric_hub.py:51]:
```suggestion
    def _enquote(self, x: str) -> Optional[str]:
```

File: jobs/kpi-forecasting/Dockerfile
Comment (bochocki) [reply jobs/kpi-forecasting/Dockerfile]: Thank you for the reference! I added `--no-cache-dir` to the pip commands.

File: jobs/kpi-forecasting/Dockerfile
Comment (bochocki) [reply jobs/kpi-forecasting/Dockerfile]: Thank you, it makes sense now that you point it out but I haven't had enough experience with building Docker containers to catch that. I've made this change.

File: jobs/kpi-forecasting/.gitignore
Comment (bochocki) [reply jobs/kpi-forecasting/.gitignore]: done.

File: jobs/kpi-forecasting/README.md
Comment (bochocki) [reply jobs/kpi-forecasting/README.md:47]: Done. By default conda adds environments to a common directory (`~/Users/username/conda/envs`) but it's easy enough to add `envs` to `.gitignore` in case we choose to store environments in the repo in the future.

File: jobs/kpi-forecasting/kpi-forecasting/kpi_forecasting.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/kpi_forecasting.py:14]: That's a great question. I considered merging them, but I think that keeping them separate is helpful if the forecasting code is being run outside of a Docker context. For example, when developing locally in a Jupyter notebook, the code can instead look like:
```python
    config = YAML(filepath="local_config.yml").data
```

If it's okay, I'd prefer not to make this change for now.

File: jobs/kpi-forecasting/kpi-forecasting/metric_hub.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/metric_hub.py:32]: Good call. I'm now coercing these args to dates using `pd.to_datetime`. Pandas will undoubtedly catch more edge cases than I could write out.

File: jobs/kpi-forecasting/kpi-forecasting/metric_hub.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/metric_hub.py]: This is a great idea. I initially wrote this method because when `end_date is None` I was setting `end_string = "CURRENT_DATE()"`, and I didn't want that string to include quotes. But I'm now simply setting `end_date` to the current UTC date, so I can remove the `_enquote` method.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:32]: Nope! Removed.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py]: Removed.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py]: Removed.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:46]: There are several of these `TODO` throughout the code. They're still needed; there's a "legacy" BQ data model which will be deprecated once the new data model is in place. The code is mostly written to support what we expect the new data model to look like. But there are several methods in the code (usually of the format `_method_name_legacy`) that exist just to re-capture the legacy data format until the transition is complete.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:53]: I made an effort to explain this in the PR description, but maybe that's not the best place for the explanation. I also added the following description to the readme:

> There is [an ongoing project](https://docs.google.com/document/d/18esfJraogzUf1gbZv25vgXkHigCLefazyvzly9s-1k0/edit) to re-model the way we store forecast data in Big Query. One key change in this refactor is to format output data to be as close as possible to what we expect the new data model to look like, while also preserving the legacy data model for the time being. Methods related to legacy outputs are clearly marked in the code (usually including legacy in the method name) and are commented with TODO items that indicate which code paths can be removed once the transition has been made.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:64]: It is possible but unlikely for now. This code path will be deleted in the near-future when the new data model is complete.

I made the parsing more robust so that it handles the case that you mention.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:106]: This is expected behavior in the "legacy" framework -- mobile and desktop forecasts returned different numbers of columns (example `"regressor_00"` [here](https://github.com/mozilla/docker-etl/blob/cdc52b62b4ef7150597733ba03829bf050b28e99/jobs/kpi-forecasting/kpi-forecasting/Utils/FitForecast.py#L33)). Most of these columns are not being used (hence the project to revamp the data model), but I left the logic in place for consistency during the transition.

File: jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py:25]: I think it's okay. The [docs](https://docs.python.org/3/library/exceptions.html#ValueError) say:
> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as [IndexError](https://docs.python.org/3/library/exceptions.html#IndexError).

File: jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py:22]: 💯 clever! done.

File: jobs/kpi-forecasting/requirements.txt
Comment (bochocki) [reply jobs/kpi-forecasting/requirements.txt:16]: As far as I know they are... I started from a fresh environment and only installed the packages required to get the code running.

File: jobs/kpi-forecasting/kpi-forecasting/metric_hub.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/metric_hub.py]: I removed this method.

File: jobs/kpi-forecasting/ci_job.yaml
Comment (bochocki) [reply jobs/kpi-forecasting/ci_job.yaml]: I don't understand why, but the `build-docker-etl` check is still failing after updating from master and making this change.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (kik-kik) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:53]: If this is specific to the code, I'd include that information with the code. It's much easier to read and understand, I may look at the code without knowing that README has code specific comments.

File: jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py
Comment (kik-kik) [reply jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py:23]: That's fair, I'd maybe make the message a little bit more user friendly and include in it what the expected values are.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (kik-kik) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:46]: That's okay, just wanted to make sure this is actually should be in the code. One suggestion I'd have, it sounds like there might be an issue related to this TODO, if that indeed is the case, could you include the ticket number with this comment to be able to find the status of this and additional context.

File: jobs/kpi-forecasting/kpi-forecasting/kpi_forecasting.py
Comment (kik-kik) [reply jobs/kpi-forecasting/kpi-forecasting/kpi_forecasting.py:14]: of course this is fine. The modularity is good, was just unsure if really unnecessary with so little code.

**Comment by kik-kik** [comment]
More general comment @bochocki, would it make sense to add unit tests for this code?

**Comment by kik-kik** [comment]
@bochocki I made a commit to the branch to fix the `build-docker-etl` CI step. Basically the reason it was failing is that we needed to re-run `./script/update_ci_config` script to apply the changes to your `ci_job.yaml` definition to `.circleci/config.yml` file.

**Comment by bochocki** [comment]
> More general comment @bochocki, would it make sense to add unit tests for this code?

Unit tests are definitely a good idea. I wonder if it'll be okay to add those in a follow-up PR? I'll need to mock data for some of the more useful tests and I'd like to think a bit more carefully about that before diving into an implementation.

**Comment by kik-kik** [comment]
> > More general comment @bochocki, would it make sense to add unit tests for this code?
> 
> Unit tests are definitely a good idea. I wonder if it'll be okay to add those in a follow-up PR? I'll need to mock data for some of the more useful tests and I'd like to think a bit more carefully about that before diving into an implementation.

I'll leave that up to you to decide.

**Review by kik-kik — APPROVED** [review]
I assume you tested the code and it works as expected. I think it'd be great if we could add tests, but also won't block this change. Please resolve any open conversations prior to merging.

File: jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py]: Done!

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:46]: I added a link to the Jira ticket instead of a link to the proposal doc.

File: jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py
Comment (bochocki) [reply jobs/kpi-forecasting/kpi-forecasting/models/prophet_forecast.py:53]: I removed this from the README in favor of including a link to the Jira ticket in the comments.

**Pull request merged**
